### PR TITLE
Build System: Use PACKAGE_TARNAME in NSIS script

### LINKED ISSUE #7265: setup.nsi.in has "bitcoin" and "Bitcoin Core" hardcoded a lot
https://github.com/bitcoin/bitcoin/blob/master/share/setup.nsi.in has the strings "bitcoin" and "Bitcoin Core" hardcoded in many places, even though those strings are available via autoconf.  This makes it slightly more bothersome to change the name of the project, because changing the AC_INIT line in configure.ac isn't sufficient.

I don't know if this is something you want to fix in your repo, since it mostly only affects downstream projects like Namecoin.  If you are interested in fixing this, I'd be happy to submit a PR.

### PR DESCRIPTION
**Pull request by JeremyRand**
This PR replaces the hardcoded string "bitcoin" in the NSIS script with the autoconf variable PACKAGE_TARNAME; fixes https://github.com/bitcoin/bitcoin/issues/7265 .

Places where I chose not to replace:
1. bitcoin.ico wasn't replaced because it doesn't seem to be relevant to the build system and its filename never affects the end user.
2. InstallDir wasn't replaced because the current text has an uppercase B, and I'm not sure of a good way to capitalize the result of PACKAGE_TARNAME.
3. A comment in the Main Installer section wasn't replaced because comments don't ever face the end user.
4. The registry value "URL:Bitcoin" wasn't replaced for the same reason as InstallDir.
5. Startup shortcut wasn't replaced for the same reason as InstallDir.

All other appearances of "bitcoin" in the NSIS script were replaced.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Concept ACK.

**Comment by jonasschnelli** [comment]
Concept ACK.
Kicked a build to see if this works over gitian/depends build: https://bitcoin.jonasschnelli.ch/pulls/7603/

**Comment by MarcoFalke** [comment]
Looks like gitian is happy with this

File: share/setup.nsi.in [hunk 1]
```diff
@@ -20,7 +20,7 @@ SetCompressor /SOLID lzma
 !define MUI_STARTMENUPAGE_REGISTRY_KEY ${REGKEY}
 !define MUI_STARTMENUPAGE_REGISTRY_VALUENAME StartMenuGroup
 !define MUI_STARTMENUPAGE_DEFAULTFOLDER "@PACKAGE_NAME@"
-!define MUI_FINISHPAGE_RUN $INSTDIR\bitcoin-qt.exe
+!define MUI_FINISHPAGE_RUN $INSTDIR\@PACKAGE_TARNAME@-qt.exe
```
Comment (laanwj) [inline share/setup.nsi.in]: Hmm, if we're parametrizing executable names anyway I'd prefer to be able to customize the entire name, e.g. have a `@BITCOIN_QT_NAME@`, and set that to `@PACKAGE_TARNAME@-qt` somewhere higher in the build system.

There's been talk of getting rid of the `-qt` suffix for the user interface in the past, so having it hard-coded in as few places as possible would be a step forward.

File: share/setup.nsi.in [hunk 2]
```diff
@@ -73,19 +73,19 @@ ShowUninstDetails show
...
-    File @abs_top_srcdir@/release/bitcoin-cli.exe
+    File @abs_top_srcdir@/release/@PACKAGE_TARNAME@d.exe
+    File @abs_top_srcdir@/release/@PACKAGE_TARNAME@-cli.exe
     SetOutPath $INSTDIR\doc
     File /r @abs_top_srcdir@/doc\*.*
     SetOutPath $INSTDIR
     WriteRegStr HKCU "${REGKEY}\Components" Main 1
 
     # Remove old wxwidgets-based-bitcoin executable and locales:
-    Delete /REBOOTOK $INSTDIR\bitcoin.exe
```
Comment (laanwj) [inline share/setup.nsi.in]: Heh, these are from `<0.5.0`, probably not a bad time to remove them.

File: share/setup.nsi.in (hunk 1 above)
Comment (JeremyRand) [inline share/setup.nsi.in]: I agree that that would be useful.  Unfortunately my familiarity with autoconf is limited enough that I'm not fully comfortable implementing that myself.  Maybe someone else could do that as another pull request?

File: share/setup.nsi.in (hunk 2 above)
Comment (JeremyRand) [inline share/setup.nsi.in]: I'd be fine with that.  Should that be done as part of this pull request?

Comment (MarcoFalke) [inline share/setup.nsi.in]: Sure, go ahead.

File: share/setup.nsi.in (hunk 1 above)
Comment (laanwj) [inline share/setup.nsi.in]: Fine with me.

**Comment by laanwj** [comment]
@theuni I suppose this is ok with you?

**Comment by JeremyRand** [comment]
I just rebased against latest master, and added a 2nd commit which removes the wxwidgets stuff as @laanwj and @MarcoFalke requested.

**Comment by theuni** [comment]
Ooh neat, I'm assigned. I guess I should take that as a hint and fix my mail filters so I'm not missing pings :)

Agree with @laanwj wrt avoiding hard-coding filenames. The names are already set in the top Makefile as @BITCOIN_QT_BIN@ and @BITCOIN_WIN_INSTALLER@, we just need to modify BITCOIN_QT_BIN to not include a path. I'll work up a quick change for that.

**Comment by theuni** [comment]
@JeremyRand Can you please take the top two commits from https://github.com/theuni/bitcoin/tree/7603 ? You can squash the NSIS one into your own. After that, I'm good with this.

We can't use BITCOIN_QT_BIN/BITCOIN_WIN_INSTALLER as I said above because they're not yet defined. I added new vars and used them in one place to ensure that everything stays synced up.

**Comment by JeremyRand** [comment]
@theuni Is it okay if I modify the 4 lines at https://github.com/theuni/bitcoin/commit/411a71f57e075c7d0a7c79bf1f7c7d249d5145da#diff-67e997bcfdac55191033d57a16d1408aR17 to use PACKAGE_TARNAME instead of hardcoding the "bitcoin" substring on those lines?

**Comment by theuni** [comment]
@JeremyRand I'd prefer not. The naming there should be the canonical definition, so it really shouldn't depend on anything else.

**Comment by JeremyRand** [comment]
@theuni Hmm, in the use cases for changing these variables that I personally deal with (i.e. Namecoin) it's pretty much a given that when PACKAGE_TARNAME is changed, the binary filenames are going to change accordingly as well.  Is there a use case I'm not seeing, where it's likely that PACKAGE_TARNAME and the prefix of the binary filenames will be different?  (I admit that I'm not particularly familiar with the variety of customizations that people might make to the build scripts.)

**Comment by laanwj** [comment]
Well the idea would be to keep as much flexibility as possible, so to be able to configure various names separately from each other.
I understand that you ideally want to keep your patch as small as possible, but if changing the executable names is at least centralized in one place that doesn't blow it up much.

**Comment by JeremyRand** [comment]
@laanwj Yes, agreed that flexibility is important here.  However, is there a loss in flexibility if a forked project has to change "$PACKAGE_TARNAME" to something else in the four binary name variables, rather than changing "bitcoin" to something else in the same variables?  I'm definitely not arguing that those four variables should be removed, I just think it makes sense for their default values to be dependent on PACKAGE_TARNAME since that seems to be a common configuration.  Any forked project that wants the binary filename prefixes to be something other than PACKAGE_TARNAME can change them just as easily as changing them away from being "bitcoin".  So I don't see a loss of flexibility -- am I missing something?

**Comment by MarcoFalke** [comment]
The loss in flexibility appears when someone changes PACKAGE_TARNAME and is surprised all the binaries changed their name, too. So, has to go back and adjust the names of the binaries...

I think 644e6de8f3e8a86022a2daa4666bb28dcd5a4e24 is an uncontroversial improvement (after squash), which can be merged right now. If you really want to refactor PACKAGE_TARNAME, I'd prefer another pull for this.

**Comment by theuni** [comment]
I agree with @MarcoFalke.

@JeremyRand While it's obviously nice if our changes make life easier for other projects, our first priority has to be to our own.

There are plenty of scenarios where the 2 values could become out of sync, and forks may or may not wish to switch when we do. So inevitably, there's going to be some patching down the road. I'd rather not create something that appears to be a contract to downstreams in the process.

**Comment by JeremyRand** [comment]
@theuni @MarcoFalke Okay, sounds good, you've convinced me.  I'll add the suggested commits to this PR when I have a few minutes (I'm traveling today).

**Comment by JeremyRand** [comment]
There we go.  Hopefully I did the requested squash correctly.

**Comment by MarcoFalke** [comment]
utACK 0528e30

**Comment by laanwj** [comment]
Looks much better now!
utACK
